### PR TITLE
Code cleanup - switch to central parameters file

### PR DESCRIPTION
**Pull request by mnirkko**
A single file 'tellie.cfg' now sets all the necessary parameters to avoid redundancy. The parameters are read out by 'parameters.py' using the package ConfigParser, all scripts that import parameters can use them. Major change, successfully tested at Sussex test box. Log attached:
[testing_2017_03_17.txt](https://github.com/snoplus/tellie/files/851926/testing_2017_03_17.txt)

### CODE REVIEW

File: core/serial_command.py [hunk 1]
```diff
@@ -275,12 +228,12 @@ def trigger_single(self):
         if self._firing is True:
             raise tellie_exception.TellieException("Cannot fire, already in firing mode")
         #if self._channel <= 56: #up to box 7                                                               
-        #    cmd = _cmd_fire_single_lower
+        #    cmd = p._cmd_fire_single_lower
         #else:
-        #    cmd = _cmd_fire_single_upper
-        self._send_command(_cmd_fire_ext_trig, False)
+        #    cmd = p._cmd_fire_single_upper
```
Comment (EdLeming) [inline core/serial_command.py]: We should remove these commented out lines

File: core/serial_command.py [hunk 2]
```diff
@@ -362,30 +315,30 @@ def fire_continuous(self):
...
         if self._firing is True:
             raise tellie_exception.TellieException("Cannot fire, already in firing mode")
-        self._send_command(_cmd_fire_continuous, False)
+        self._send_command(p._cmd_fire_continuous, False)
         self._firing = True
         self._force_setting = False
 
-    def read_buffer(self, n=100):
+    def read_buffer(self, n=p._read_bytes):
         return self._serial.read(n)
```
Comment (EdLeming) [inline core/serial_command.py]: We don't seem to use this function, but I think we should. Generally we call self._serial.read(n) directly with n set to 100. However, If we want to change the default p._read_bytes, this means we'll have to update every single instance. We should be using this method instead

File: core/serial_command.py [hunk 3]
```diff
@@ -362,30 +315,30 @@ def fire_continuous(self):
...
 
     def stop(self):
         """Stop firing tellie"""
         self.logger.debug("Stop firing!")
         #Disable external trigger before we do anything
         self.disable_external_trigger()
-        self._send_command(_cmd_stop, False)
-        buffer_contents = self._serial.read(100)
+        self._send_command(p._cmd_stop, False)
+        buffer_contents = self._serial.read(p._read_bytes)
```
Comment (EdLeming) [inline core/serial_command.py]: here's an example of what I've mentioned above. We should use the read buffer method instead !

File: core/serial_command.py [hunk 4]
```diff
@@ -398,24 +351,24 @@ def read_pin(self, channel=None, timeout=2.0, final=True):
             else:
                 self.select_channel(channel)
             if self._channel[0] <= 56: #up to box 7
-                #cmd = _cmd_read_average_lower
+                #cmd = p._cmd_read_average_lower
```
Comment (EdLeming) [inline core/serial_command.py]: delete?

File: core/serial_command.py [hunk 5]
```diff
@@ -398,24 +351,24 @@ def read_pin(self, channel=None, timeout=2.0, final=True):
...
                 print "read!"
-                cmd = _cmd_read_single_lower
+                cmd = p._cmd_read_single_lower
             else:
-                #cmd = _cmd_read_average_upper
+                #cmd = p._cmd_read_average_upper
```
Comment (EdLeming) [inline core/serial_command.py]: delete?

File: core/serial_command.py [hunk 6]
```diff
@@ -532,12 +485,12 @@ def select_channels(self, channels):
...
         self.clear_channel()
-        command = _cmd_channel_select_many_start
+        command = p._cmd_channel_select_many_start
         for channel in channels:
             print channel
             command += chr(channel)
-        command += _cmd_channel_select_many_end
-        buffer_check = "B"+str((int(channels[0])-1)/8+1)+_cmd_channel_select_many_end
+        command += p._cmd_channel_select_many_end
+        buffer_check = p._cmd_disable_ext_trig+str((int(channels[0])-1)/8+1)+_cmd_channel_select_many_end
```
Comment (EdLeming) [inline core/serial_command.py]: I think _cmd_channel_select... should have a p first?

File: core/serial_command.py [hunk 7]
```diff
@@ -810,33 +763,33 @@ def select_channel(self, channel):
 
 def command_select_channel(par):
     """Get the command to select a single channel"""
-    command = _cmd_channel_select_single_start+chr(par)+_cmd_channel_select_single_end
-    buffer_check = "B"+str((int(par)-1)/8+1)+_cmd_channel_select_single_end
+    command = p._cmd_channel_select_single_start+chr(par)+_cmd_channel_select_single_end
+    buffer_check = p._cmd_disable_ext_trig+str((int(par)-1)/8+1)+_cmd_channel_select_single_end
```
Comment (EdLeming) [inline core/serial_command.py]: again, this one (and the line above), I think you're missing a couple of p's. _cmd_channel_select... isn't currently being imported from the parameters file

File: core/serial_command.py [hunk 8]
```diff
@@ -810,33 +763,33 @@ def select_channel(self, channel):
...
         raise tellie_exception.TellieException("Invalid pulse height: %s" % par)
-    hi = par >> 8
-    lo = par & 255
+    hi = par >> 8   # binary right shift
+    lo = par & 255  # binary AND operator
     command = [_cmd_pulse_height_hi+chr(hi)]
     command+= [_cmd_pulse_height_lo+chr(lo)]
     command+= [_cmd_pulse_height_end]
-    buffer_check = _cmd_pulse_height_hi + _cmd_pulse_height_lo + _cmd_pulse_height_end
+    buffer_check = p._cmd_pulse_height_hi + _cmd_pulse_height_lo + _cmd_pulse_height_end
```
Comment (EdLeming) [inline core/serial_command.py]: missing p's?

File: core/serial_command.py [hunk 9]
```diff
@@ -810,33 +763,33 @@ def select_channel(self, channel):
...
     if par > _max_pulse_width or par < 0:
         raise tellie_exception.TellieException("Invalid pulse width: %s %s %s" % (par, _max_pulse_width, par>_max_pulse_width))
-    hi = par >> 8
-    lo = par & 255
+    hi = par >> 8   # binary right shift
+    lo = par & 255  # binary AND operator
     command = [_cmd_pulse_width_hi+chr(hi)]
     command+= [_cmd_pulse_width_lo+chr(lo)+_cmd_pulse_width_end]
-    buffer_check = _cmd_pulse_width_hi + _cmd_pulse_width_lo + _cmd_pulse_width_end
+    buffer_check = p._cmd_pulse_width_hi + _cmd_pulse_width_lo + _cmd_pulse_width_end
```
Comment (EdLeming) [inline core/serial_command.py]: this method is missing p's in a number of places.

File: core/serial_command.py [hunk 10]
```diff
@@ -845,12 +798,12 @@ def command_pulse_number(par):
...
         raise tellie_exception.TellieException("Invalid pulse number: %s" % (par))
     par = int(par)
-    adjusted, actual_par, hi, lo = parameters.pulse_number(par)
+    adjusted, actual_par, hi, lo = p.pulse_number(par)
     if adjusted is True:
         raise tellie_exception.TellieException("Invalid pulse number: %s" % (par))
     command = [_cmd_pulse_number_hi+chr(hi)]
     command+= [_cmd_pulse_number_lo+chr(lo)]
-    buffer_check = _cmd_pulse_number_hi + _cmd_pulse_number_lo
+    buffer_check = p._cmd_pulse_number_hi + _cmd_pulse_number_lo
```
Comment (EdLeming) [inline core/serial_command.py]: missing p after + ?

File: core/serial_command.py [hunk 11]
```diff
@@ -862,7 +815,7 @@ def command_pulse_delay(par):
     us = int((par-ms)*250)
     command = [_cmd_pulse_delay+chr(ms)]
```
Comment (EdLeming) [inline core/serial_command.py]: missing p?

**Comment by EdLeming** [comment]
So I've started to look through this, but before I get further I think we should think about stream lining the tellie server code. Rather than copy / pasting the whole serial_command.py code, we should simply import the class and load it into the server - making keeping track of the code significantly easier. 

Could you try to do so and run some tests? This would be a significant improvement and I think it's one we really should make.

File: core/serial_command.py (hunk 1 above)
Comment (mnirkko) [inline core/serial_command.py]: done

File: core/serial_command.py (hunk 2 above)
Comment (mnirkko) [inline core/serial_command.py]: Done, replaced all other calls to _serial_read(n) with read_buffer(), i.e. using the default value p._read_bytes

File: core/serial_command.py (hunk 3 above)
Comment (mnirkko) [inline core/serial_command.py]: see above

File: core/serial_command.py (hunk 4 above)
Comment (mnirkko) [inline core/serial_command.py]: jep

File: core/serial_command.py (hunk 5 above)
Comment (mnirkko) [inline core/serial_command.py]: done

File: core/serial_command.py (hunk 6 above)
Comment (mnirkko) [inline core/serial_command.py]: Correct, well spotted. No idea why this didn't crash, maybe the function wasn't being called.

File: core/serial_command.py (hunk 7 above)
Comment (mnirkko) [inline core/serial_command.py]: done

File: core/serial_command.py (hunk 8 above)
Comment (mnirkko) [inline core/serial_command.py]: done

File: core/serial_command.py (hunk 9 above)
Comment (mnirkko) [inline core/serial_command.py]: done

File: core/serial_command.py (hunk 10 above)
Comment (mnirkko) [inline core/serial_command.py]: done

File: core/serial_command.py (hunk 11 above)
Comment (mnirkko) [inline core/serial_command.py]: Done, made sure via search function that all _cmd_ calls have a "p." in front of them.

**Comment by mnirkko** [comment]
Thanks Ed. I've already started commenting out duplicate functions in the server code that were clearly the same. For an entire class like SerialCommand, that may be more tricky. In the worst case I'll just do a "server" flag (boolean) that does X if true and Y if false. That can certainly be in one place, and it beats preprocessing flags. ;)

**Comment by EdLeming** [comment]
OK, so I was thinking about this on my walk in this morning. I think the solution is to simply delete serial_command.py. We can import everything all the test scripts need from the server instead (as the class is duplicated there). What do you think?

**Comment by mnirkko** [comment]
Yeah, that sounds doable. I'm worried it might make the tellie_server.py 
a bit heavy, but probably not much more than it already is (currently 
both scripts at 900 lines). It's worth a try.

**Comment by EdLeming** [comment]
Hey, @mnirkko. I think this looks good! One thing to consider really is the size of these changes. Are we certain we have tested to the point that we're happy that this can be run underground?

**Comment by mnirkko** [comment]
I understand this point well. However, I am not certain how to test it 
better than checking the output underground and seeing whether we're 
happy with it. At least the parameters file will need some minor changes 
(select correct USB port + log file locations), but apart from that I 
"expect" it to work (didn't change the juicy parts, just moved them around).

**Pull request merged**